pinot-core/src/main/java/org/apache/pinot/core/transport/QueryResponse.java

```diff
@@ -49,12 +50,12 @@ enum Status {
   /**
    * Returns the current server responses without blocking.
    */
-  Map<ServerRoutingInstance, ServerResponse> getCurrentResponses();
+  Map<ServerRoutingInstance, List<ServerResponse>> getCurrentResponses();
```

Comment:
I don't think the code change in this PR is backward compatible. We definitely need to keep the existing signatures as is. Can we introduce new signatures and mark the existing ones `Deprecated`, so that the old signatures can be cleaned up in the next official release?

Reply:
I fully recognize that I may be missing some important details here, but can you help me understand the implications of changing this interface?  I see that the interface `QueryResponse` and its only implementation `AsyncQueryResponse` are defined in the `org.apache.pinot.core.transport` package, but all usages are isolated to `org.apache.pinot.broker`.  

I agree that the change needs to be backward compatible with the data sent from Pinot servers to the brokers so that mismatched versions of code running on brokers Vs servers can interoperate during upgrade. But I'm not yet certain that this interface needs to remain unchanged in order to accomplish that.

Reply:
@jackjlli I believe I've found a way to maintain compatibility such that brokers with this version could roll out and communicate with servers on a prior version, while still gathering the information required.  This is done through borrow a single digit of the request ID (the least significant digit, such that it does not increase the probability of a broker ID hash collision) in order to use that bit to identify if a query pertained to a REALTIME or OFFLINE table.

Because in this version servers will only receive 1 or 2 queries, always targeting the same table "base" (i.e. foo of foo_OFFLINE), we can infer from the table type REALTIME or OFFLINE which request the server response pertains to. This system will not work in the future when a server will return responses for many different tables, but as soon as we have server versions deployed which always return the table name in the dataTable metadata, we can fully dismantle this requestId hijacking. Using the request ID to convey table type only needs to exist for the moments (or hours, depending on size of cluster) where brokers are expecting table name in dataTable metadata, but servers are not yet running the version which does so.

I feel like this provides a good bridge to be able to move forward with subsequent Logical Table implementation.

Reply:
These changes specifically can be viewed like so (a subset of overall commits): https://github.com/apache/pinot/pull/13742/files/2ffd65d8f869f5f707af708f1680cbb45fb3a2ea..734189bc7da58705f6090102abbec76758ab84ee

Reply:
Also wanted to add that the way the server responds is not changing such that the servers return a list and therefore the brokers need to change what they expect. It's only that all responses from the same server are grouped as a list upon receipt within the brokers.